birdhouse/optional-components/test-cowbird-jupyter-access/docker-compose-extra.yml

```diff
@@ -0,0 +1,27 @@
+version: "3.4"
+services:
+  jupyterhub:
+    environment:
+      TEST_COWBIRD_JUPYTERHUB_USERNAME: ${TEST_COWBIRD_JUPYTERHUB_USERNAME}
```

Comment:
So this is the only change to other service (`jupyterhub`) than add the new service (`test-cowbird-jupyter-access`) just to be able to run the script once.

The `jupyterhub` service modification is not even necessary for Jenkins since Jenkins do not run on the PAVICS  platform.

So I do not see why this bootstrap script can not be called by the bootstrap entrypoint instead of being in a new component that is not really a service.

Now I am a bit lost as to why Jenkins works properly since none of all the extra changes to `jupyterhub` component is ever taken by Jenkins.  Aka, all the extra configs in `JUPYTERHUB_CONFIG_OVERRIDE` is not seen at all by Jenkins.

Reply:
> Jenkins do not run on the PAVICS platform.

Just to clarify on this point, Jenkins do test against the PAVICS platform, but the Jupyter env runtime environment that the test notebooks runs on Jenkins is not spawned by the JupyterHub component, so any changes to the JupyterHub component is not seen/tested by Jenkins.

So if the test notebook works under Jenkins, it means the changes to the JupyterHub component might not be required.  Maybe the new bootstrap script is all that is required, no need for JupyterHub config extra config changes.

Please either review the notebook or the necessity of JupyterHub extra configs.

Reply:
The way my test notebooks work is this :

- The [main notebook](https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/blob/cowbird-jupyter-e2e-test/notebooks-auth/test_cowbird_jupyter.ipynb) that is executed first is indeed not spawned by the JupyterHub component, but it does use the JupyterHub component. I highlighted the cells that use JuptyerHub [here](https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/131/files#diff-8f663cf7d0e2fa8d8be62174e0b10b59f24e8486189d0997e9748c4dfc5e2cacR142-R196). It checks first that the test user exists on JupyterHub, gets its token, and then uses the `jhub-client` package to spawn a new JupyterLab with the usual DockerSpawner. So, the JupyterHub is indeed tested here.

- The [second notebook](https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/blob/cowbird-jupyter-e2e-test/notebooks-auth/resources/user_test_cowbird_jupyter.ipynb) is the notebook that is tested against the PAVICS platform, on the JupyterHub environment. 

So, the changes to the JupyterHub component are uses to add role permissions to be able to do the operations on the main notebook (creating a user, getting a user token), and the environment variables added to the DockerSpawner environment are used by the second notebook's test.

Reply:
@tlvu 
Sadly, @ChaamC as already spent ~2x the amount of time originally planned for this month to complete this test validation, given how complex it can get with the double nesting of notebook executions, Jupyter API/kernel interactions, combined with Magpie/Cowbird permissions. Given the month is nearing the end, he will not have further time to refactor this.

Since there are little to no validation tests of the Jupyter/Cowbird interactions at the moment, I would like to suggest moving forward with this first implementation, and consider refactors if really judged necessary at a later point. IMO, given that part of the configuration must be done *before* the stack is started (Jupyter config/token), and other *after* (test user/permission setup), I believe that splitting them up into bootstrap/optional-component will only make this validation even more complicated than it already is. It will be much easier for someone to work on this feature if all configurations and scripts are combined under the same optional-component and executed roughly at the same time on startup.

I would like also to point out that restarting the stack should not restart that service each time if it is set to `restart: "no"`. It will also not report that it is down, but that it did a successful "Exit 0". The only thing that @ChaamC could do as final check is that conflict conditions that could make the script fail are handled gracefully in case the service is forcefully restarted.

Reply:
@fmigneault @ChaamC I wasn't against leaving it as-is. I realized it's a bit late to change the design.  I was just inquiring if any side-effects have been anticipated (ex: would re-run cause errors on force restart, anything else ...)

Reply:
So, if the `test-cowbird-jupyter-access` container from the optional-component is restarted, its script will get reexecuted, but the first thing the script does, apart from some variable assignations and some checks, is to try to create the test user on Magpie. If the user already exists, it raises an error and stops. Considering this, I don't think there should be any side effect if the stack is restarted, as it won't do anything if the test user already exists.

The effects brought by the script are :
- the creation of a test user Magpie
- creating some files in the test user's own workspace
- creating public and user-specific WPS outputs files in `/data/wps_outputs`